### PR TITLE
[SCB-483] Springmvc rest vertx file download

### PR DESCRIPTION
**Pull request by wujimin**
springmvc over rest vertx server support file download.
image

### CODE REVIEW

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java [hunk 1]
```diff
@@ -220,10 +216,38 @@ protected void sendResponse(Response response) throws Exception {
...
+      responseEx.flushBuffer();
+    } catch (IOException flushException) {
+      LOGGER.error("Failed to flush rest response, operation:{}.",
+          invocation.getMicroserviceQualifiedName(),
+          flushException);
+    }
+
+    requestEx.getAsyncContext().complete();
+    // if failed to locate path, then will not create invocation
+    // TODO: statistics this case
```
Comment (WillemJiang) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: We may also need to count the error and exceptions.

File: foundations/foundation-vertx/src/main/java/org/apache/servicecomb/foundation/vertx/stream/InputStreamToReadStream.java [hunk 2]
```diff
@@ -34,7 +34,7 @@
 public class InputStreamToReadStream implements ReadStream<Buffer> {
   private static final Logger LOGGER = LoggerFactory.getLogger(InputStreamToReadStream.class);
 
-  public static final int DEFAULT_READ_BUFFER_SIZE = 8192;
+  public static final int DEFAULT_READ_BUFFER_SIZE = 1024 * 1024;
```
Comment (WillemJiang) [inline foundations/foundation-vertx/src/main/java/org/apache/servicecomb/foundation/vertx/stream/InputStreamToReadStream.java]: We may need to mention it in the release note.

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java (hunk 1 above)
Comment (wujimin) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: traced by https://issues.apache.org/jira/browse/SCB-500

File: foundations/foundation-vertx/src/main/java/org/apache/servicecomb/foundation/vertx/stream/InputStreamToReadStream.java (hunk 2 above)
Comment (wujimin) [reply foundations/foundation-vertx/src/main/java/org/apache/servicecomb/foundation/vertx/stream/InputStreamToReadStream.java]: only a internal buf size, customers not perceive this?

**Comment by wujimin** [comment]
add a new commit: "set correct content-type for download"

File: demo/demo-springmvc/springmvc-server/src/main/java/org/apache/servicecomb/demo/springmvc/server/DownloadSchema.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+
+@RestSchema(schemaId = "download")
+@RequestMapping(path = "/download", produces = MediaType.APPLICATION_OCTET_STREAM_VALUE)
+public class DownloadSchema {
+  @GetMapping(path = "/file")
```
Comment (liubao68) [inline demo/demo-springmvc/springmvc-server/src/main/java/org/apache/servicecomb/demo/springmvc/server/DownloadSchema.java]: Are you going to provide a way to invoke download service by RestTemplate or RPC style?

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/filter/inner/ServerRestArgsFilter.java [hunk 4]
```diff
@@ -50,7 +54,7 @@ public Response afterReceiveRequest(Invocation invocation, HttpServletRequestEx
   }
 
   @Override
-  public void beforeSendResponse(Invocation invocation, HttpServletResponseEx responseEx) {
+  public CompletableFuture<Void> asyncBeforeSendResponse(Invocation invocation, HttpServletResponseEx responseEx) {
```
Comment (liubao68) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/filter/inner/ServerRestArgsFilter.java]: I think  beforeSendResponseAsync is a better name

Comment (wujimin) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/filter/inner/ServerRestArgsFilter.java]: done.

**Pull request merged**